plugins/modem-manager/fu-sahara-loader.c

```diff
@@ -0,0 +1,644 @@
+			dev = (void *)ptr;
+
+			/* Devices in QDL mode always have these vid:pid */
+			if (dev->idVendor != 0x05c6 || dev->idProduct != 0x9008) {
+				g_set_error(error,
+					G_IO_ERROR,
+					G_IO_ERROR_FAILED,
+					"Wrong device and/or vendor id: 0x%04x 0x%04x",
+					dev->idVendor, dev->idProduct);
+				return FALSE;
+			}
+		} else if (hdr->bDescriptorType == USB_DT_INTERFACE && hdr->bLength == USB_DT_INTERFACE_SIZE) {
+			ifc = (void *)ptr;
+
+			ifclass = ifc->bInterfaceClass;
+			ifsubclass = ifc->bInterfaceSubClass;
+			ifnum = ifc->bInterfaceNumber;
+
+			self->in_ep = -1;
+			self->out_ep = -1;
+		} else if (hdr->bDescriptorType == USB_DT_ENDPOINT && hdr->bLength == USB_DT_ENDPOINT_SIZE ) {
+			ept = (void *)ptr;
+
+			if (ifclass == 0xff && ifsubclass == 0xff) {
+				if ((ept->bmAttributes & USB_ENDPOINT_XFERTYPE_MASK) == USB_ENDPOINT_XFER_BULK) {
+					if (ept->bEndpointAddress & USB_ENDPOINT_DIR_MASK) {
+						self->in_ep = ept->bEndpointAddress;
+						self->in_maxpktsize = ept->wMaxPacketSize;
+					} else {
+						self->out_ep = ept->bEndpointAddress;
+						self->out_maxpktsize = ept->wMaxPacketSize;
+					}
+
+					if (self->in_ep != -1 && self->out_ep != -1) {
+						self->ifnum = ifnum;
+						return TRUE;
+					}
+				}
+			}
+		}
+
+		ptr += hdr->bLength;
+	}
+
+	g_set_error(error,
+		G_IO_ERROR,
+		G_IO_ERROR_FAILED,
+		"Couldn't find USB interface and endpoint for Sahara protocol");
+
+	return FALSE;
+}
+
+static gboolean
+fu_sahara_loader_claim_interface(FuSaharaLoader *self, GError **error)
+{
+	int res;
+	struct usbdevfs_ioctl cmd;
+
+	cmd.ifno = self->ifnum;
+	cmd.ioctl_code = USBDEVFS_DISCONNECT;
+	cmd.data = NULL;
+
+	res = ioctl(fu_io_channel_unix_get_fd(self->io_channel), USBDEVFS_IOCTL, &cmd);
+	if (res && errno != ENODATA) {
+		g_set_error(error,
+			G_IO_ERROR,
+			G_IO_ERROR_FAILED,
+			"Failed to disconnect kernel driver");
+		return FALSE;
+	}
+
+	res = ioctl(fu_io_channel_unix_get_fd(self->io_channel), USBDEVFS_CLAIMINTERFACE, &self->ifnum);
+	if (res < 0) {
+		g_set_error(error,
+			G_IO_ERROR,
+			G_IO_ERROR_FAILED,
+			"Failed to claim USB Sahara interface");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+gboolean
+fu_sahara_loader_open(FuSaharaLoader *self, GError **error)
+{
+	const gchar *devname;
+	g_autoptr(GUdevClient) client = NULL;
+	g_autoptr(GUdevDevice) device = NULL;
+	g_autoptr(GByteArray) usb_desc = NULL;
+
+	g_debug("opening sahara port...");
+
+	client = g_udev_client_new(NULL);
+	device = g_udev_client_query_by_sysfs_path(client, self->sysfs_path);
+	if (device == NULL) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_FAILED,
+			    "Failed to get device from the path: %s",
+			    self->sysfs_path);
+		return FALSE;
+	}
+
+	devname = g_udev_device_get_property(device, "DEVNAME");
+	if (devname == NULL) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_FAILED,
+			    "Failed to get the devname for the device: %s",
+			    self->sysfs_path);
+		return FALSE;
+	}
```

Comment:
Hmm, so you want to go from a self->sysfs_path to a GUsbDevice? I don't think reading from the device node is the best way to do this.